### DISCUSSION TITLE
Seeding the database in test environmnet

### DISCUSSION THREAD

**Discussion by glothos** [tada5hi/typeorm-extension#209] [Q&A] [answered]
Hey, there. Thanks for this awesome lib, saved me some time.

I'm having a issue when trying to seed the database in test environment. I'm probably missing something.
Here's the datasource:
`src/database/data-source.ts`
```ts
// this is only for testing, we inject configs via nest decorators
import { DataSource, DataSourceOptions } from 'typeorm';
import * as path from 'path';
import { SeederOptions } from 'typeorm-extension';
import UserSeeder from './seeds/user.seeder';
import userFactory from './factories/user.factory';

const options: DataSourceOptions & SeederOptions = {
  type: 'postgres',
  host: 'localhost',
  port: 5433,
  username: 'postgres',
  password: 'postgres',
  database: 'test_postgres',
  synchronize: true,
  entities: [path.join(__dirname, '/../**/*.entity{.ts,.js}')],
  migrations: [],
  seeds: [UserSeeder],
  factories: [userFactory],
};

export const dataSource = new DataSource(options);

```
Here's the seeder factory:
`src/database/factories/user.factory.ts`
```ts
import { User } from 'src/users/entities/user.entity';
import { Role } from 'src/core/models/role';
import { setSeederFactory } from 'typeorm-extension';

export default setSeederFactory(User, (faker) => {
  const user = new User();
  user.firstName = faker.name.firstName('male');
  user.lastName = faker.name.lastName('male');
  user.email = faker.internet.email(user.firstName, user.lastName);
  user.passwordHash = faker.random.alphaNumeric(16);
  user.gender = 'Male';
  user.birthDate = new Date();
  user.state = faker.address.state();
  user.city = faker.address.city();
  user.neighborhood = faker.address.city();
  user.role = Role.USER;

  return user;
});
```
Here's the Seeder class:
`src/database/seeds/user.seeder.ts`
```ts
import { Seeder, SeederFactoryManager } from 'typeorm-extension';
import { DataSource } from 'typeorm';
import { User } from 'src/users/entities/user.entity';

export default class UserSeeder implements Seeder {
  public async run(
    dataSource: DataSource,
    factoryManager: SeederFactoryManager,
  ): Promise<any> {
    const userFactory = factoryManager.get(User);
    await userFactory.saveMany(5);
  }
}
```
and here's the execution of seeder:
```ts
import UserSeeder from 'src/database/seeds/user.seeder';
import { User } from 'src/users/entities/user.entity';
import { DataSource, DataSourceOptions } from 'typeorm';
import {
  createDatabase,
  dropDatabase,
  runSeeder,
  runSeeders,
} from 'typeorm-extension';
import UserFactory from 'src/database/factories/user.factory';
import { dataSource } from 'src/database/data-source';

export const prepareDatabase = async (connection: any): Promise<void> => {
  await dropDatabase({
    options: connection as DataSourceOptions,
    ifExist: true,
  });
  await createDatabase({
    options: connection as DataSourceOptions,
    ifNotExist: true,
  });
  await dataSource.initialize();
  await runSeeders(dataSource);
  const repo = await dataSource.getRepository(User);
  const users = await repo.find();
  console.log(users);
};

```
I'm always getting this error:
`TypeORMError: No connection options were found in any orm configuration files.` in that setup.
When I use this in seeder run method:
```ts
const repository = dataSource.getRepository(User);
await repository.insert([
  {
    firstName: 'Caleb',
    lastName: 'Barrows',
    email: 'caleb.barrows@gmail.com',
  },
]);
```
it goes fine. However for the amount data I need for testing, this is really cumbersome, the seeder factory would be really useful for me in this project.
I'm using `Nest.js` in this project btw, but since all tests are ran in a jest environment I think it doesn't really matter.

**Comment by tada5hi** [comment]
Glad, you like it and happy to hear😇 .
Good catch from your side!
At the moment, the seeder factory manager tries to use the deposited singleton instance, instead of the instance provided to the runSeeder(s) method.
If no DataSource is deposited, the library attempts to locate and load the `data-source.{ts,js}` file and initialize the instance from there. But this does not seem to work in your case, so the fallback mechanism is to use the deprecated ConnectionManager from the typeorm library and then the error occurs. 
I will fix this as soon as possible. This should be solved soon.

**Reply by glothos** [reply]
Awesome. Thank you very much. I have tried a lot before opening a discussion to make sure everything was in place since I use a framework. Thank you for your quick response and help.

**Reply by glothos** [reply]
Also, it might be worth mentioning that I don't use default `TYPEORM_` prefixed env variables or `ormconfig{.json,.js}`. Since `Nest.js` provides us with typeorm module for injection and the possibility of creating a connection factory, the variables can be whatever. I used the factory because of the infrastructure server provider patterns.

**Comment by tada5hi** [comment, marked as answer]
@glothos Can you check if version `v2.1.2` is working for you?

**Reply by glothos** [reply]
You are the man! It worked perfectly. Thank you!

**Comment by tada5hi** [comment]
your welcome 😊
